### PR TITLE
[FEATURE] Add Radam optimizer

### PR DESCRIPTION
**Pull request by seujung**
## Description ##
RAdam is the large variance of the adaptive learning rates, and provide both theoretical and empirical support evidence. #899 

Modified code from https://github.com/LiyuanLucasLiu/RAdam

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by szhengac** [comment]
You need to provide clear evidence of why we should include this unpublished optimization method in gluon-nlp. For example, how does it perform in NLP tasks? What can we gain by replacing the existing optimizers with Radam in the gluon-nlp scripts? As this is an ongoing unpublished paper, you need to clearly mention the benefits in this PR.

**Comment by Ishitori** [comment]
Why adding this optimizer to GluonNLP instead of MXNet itself?

File: src/gluonnlp/optimizer/radam.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        
+        # update weight
+        # more conservative since it's an approximated value
+        if N_sma >= 5:
+            denom = sqrt(var) + self.epsilon
+            tmp = step_size * mean_hat / denom      
+            weight[:] -= step_size * mean_hat / denom
+
+        else:
+
```
[the remark this answers is not in the document]
Comment (seujung) [reply src/gluonnlp/optimizer/radam.py]: complete space removal https://github.com/dmlc/gluon-nlp/pull/903/commits/aad5c9248887230fb2a3ad642303d742da1662c1

File: src/gluonnlp/optimizer/radam.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        
+        mean_hat = mean / (1. - power(self.beta1, t))
+        
+        buffered = self.buffer[int(t % 10)]
+        if t == buffered[0]:
+            N_sma, step_size = buffered[1], buffered[2]
+        else:
+            buffered[0] = t
+            beta2_t = power(self.beta2, t)
+            N_sma_max = 2 / (1 -self. beta2) - 1
```
[the remark this answers is not in the document]
Comment (seujung) [reply src/gluonnlp/optimizer/radam.py]: complete space removal https://github.com/dmlc/gluon-nlp/pull/903/commits/aad5c9248887230fb2a3ad642303d742da1662c1

**Comment by Ishitori** [comment]
I will add RAdam to BiDAF implementation of my and let you know if it added any benefits for that particular NLP task

**Comment by seujung** [comment]
I checked the model performance in small CNN network using FashionMNIST. In this case, Radam showed a good performance. But I did not test the NLP Task. So. I will test using GluonNLP script.

**Comment by Ishitori** [comment]
The training has been completed, but unfortunately for me I don't see much of a performance improvement. The best was achieved on epoch 6 out of 12. It is still far from the paper numbers: EM = 67.7 F1 = 77.3.

```
EPOCH  6: train loss 1.9383 | batch   60 | lr 0.001 | throughtput 25.095 of samples/sec | Time per epoch 3490.67 seconds | Evaluation result  {'exact_match': 64.9763481551561, 'f1': 75.71186781682326}```

Yet, it is most probably not related to the RAdam itself.

**Comment by leezu** [comment]
@seujung would it make sense to create a PR with RAdam optimizer at https://github.com/apache/incubator-mxnet/ ? Then all MXNet based projects could profit.

**Comment by seujung** [comment]
@leezu OK. I think it's better to create a PR at mxnet. So, I'll change the RAdam to mxnet.

**Comment by szha** [comment]
@seujung closing as the PR will be made to mxnet. Let me know if you need to reopen this for any reason.

**Pull request closed without merging**